### DISCUSSION TITLE
Server implementation follow up

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#55] [General]
To allow client devs to follow each servers implementations and know what features can be implemented in clients, let's use this thread so that servers can have a post where they document what they support.
I'll update the docs from time to time but for the start and with the number of new fields that will come it's easier to have live editable data.

Clients can also post what they added to motivate servers too :)

As a starter for the new fields PR, Symfonium already support all the data except disc title, so if you add them, I'll add support for sure.

**Comment by sentriz** [comment]
for gonic:
- multi valued album artists (v0.16)
- multi valued genres (v0.16)
- transcode offsets (v0.16)

i haven't looked at the new auth stuff yet

**Reply by sentriz** [reply]
@Tolriq i just noticed that gonic is returning a concatinated list of genre strings for the "song" entity. i should be returning only 1. (and a list when we have that for `song`)

updated it now so it doesn't concat

**Reply by Tolriq** [reply]
Sorry don't understand the question :(

**Reply by sentriz** [reply]
I just noticed and issue with gonic's song response, and I know you are testing with that. but anyway it's fixed now

**Reply by Tolriq** [reply]
I only test search3 for now and did not see an issue yet it synced with your test server

**Comment by epoupon** [comment]
About LMS (https://github.com/epoupon/lms), status for version 3.42.0:

**New fields:**
* `Album` response:
  * `played`
  * `musicBrainzId`
  * `genres`
  * `artists`
  * `displayArtist`
  * `releaseTypes`
  * `moods`
  * `originalReleaseDate`
  * `isCompilation`
  * `discTitles`: discs with no subtitle are omitted
* `Child` response:
  * `played`
  * `musicBrainzId`: note this is actually the recording MBID when this response refers to a song
  * `genres`
  * `artists`
  * `displayArtist`
  * `albumArtists`
  * `displayAlbumArtist`
  * `contributors`
  * `moods`
  * `replayGain`
* `Artist` response:
  * `musicBrainzId`
  * `sortName`
  * `roles`

**Extensions:**
* Transcode offset

**Reply by epoupon** [reply]
I have put a server here with some stuff to test: https://lms-demo-os.poupon.dev   user/password is demo/demo
The difficulty for me is to find interesting free music to showcase our extensions. Maybe that would help to centralize this somewhere so we can work on the same data?

**Reply by Tolriq** [reply]
What I do for all my experiments is that I do not care about the music data just the tags, take a free album with songs that are not long, then put fake data in the tags and duplicate as many times as needed to cover all the cases. (You can also convert to some exotic file format to test the tag parser for all those cases without needing to find music in those actual formats)
Then I have a folder with all the bugged data sent by users.

**Reply by epoupon** [reply]
I thought about something like that but I had the feeling it would take me too long to edit all the relevant data we need :).
Maybe you already have some fake stuff to share?

**Reply by Tolriq** [reply]
Well no I do not have something that covers all the needs.

Here's some fake mp3 that are small with some fake data in them from ChatGPT + Stable Diffusion :p

[DJ Banana.zip](https://github.com/opensubsonic/open-subsonic-api/files/12833533/DJ.Banana.zip)

With mp3tag it's relatively easy to quickly reset and apply the tags.

For replay gain I use https://github.com/kepstin/replaygain-test-vectors/releases

**Reply by Tolriq** [reply]
@epoupon Do you think you can add some composers to the test server ? None of the albums have contributors.

**Reply by epoupon** [reply]
Ah, I selected `Demovibes 3: Pixels in sequence` because its has some composers in the track 14.
(see artist `ar-31` in album `al-2`)

**Reply by Tolriq** [reply]
Arf yes sorry, I completely forget about the default limits on search3 and it was after the limit :)

Just tested and getOpenSubsonicExtensions is not yet on the demo server? Would be nice if it was for tomorrow so I can implement everything all at once. (The transcode offset is what requires the most work on my side)

**Reply by epoupon** [reply]
No problem!
It should be fine now, I just added support of the Transcode Offset extension

**Reply by Tolriq** [reply]
The endpoint works will test the offset tomorrow. BTW I don't know if you have a setting for that, but you should probably return the json in compact format and not with tons of white space, this compress relatively well but still a lot of network ressources and CPU to uncompress for no purpose. The API is for machine not humans.

**Reply by Tolriq** [reply]
@epoupon There's a problem in your implementation the genres returned are a list of strings and not a list of ItemGenre

**Reply by Tolriq** [reply]
And @epoupon your implementation of https://opensubsonic.netlify.app/docs/endpoints/getopensubsonicextensions/ is also wrong, you return an object while it must be an array of objects.

**Reply by epoupon** [reply]
oh, I missed that point for genres. I thought it was the same as moods, which is then questionable to have different handlings for these two tags?

**Reply by Tolriq** [reply]
@epoupon it was discussed but it's the same for recordLabels, it's to allow adding ids or more properties in the future.
There will never be such need for moods.

**Reply by epoupon** [reply]
Ok, I have updated this demo instance for the extensions. Will need a little more work for the genre tags then

**Reply by epoupon** [reply]
@Tolriq : updated again!

**Reply by Tolriq** [reply]
Thanks it's for tommorrow now :)

**Reply by sentriz** [reply]
are we all based in Europe? seems like we all work at similar times

**Reply by Tolriq** [reply]
French here ;)

**Reply by sentriz** [reply]
oh nice, so is @epoupon  :ireland: :handshake: :fr:

**Reply by Tolriq** [reply]
So quickly tested the timeOffset seems to work for @epoupon but seems to be ignored for @sentriz as the media starts from the start even with the offset. Maybe you expect milliseconds ? (And this actually breaks transcoding at all since it returns flac and not requested opus)

**Reply by sentriz** [reply]
@Tolriq you need to setup transcoding on the gonic web UI. credentials are the same

**Reply by Tolriq** [reply]
@sentriz why do I need to configure?
I pass the profile in the query it should fail or have a way to know it won't transcode what I ask.
This would use user network data without knowing.

So it works when I configure but there's something wrong here from user pov and data usage risk.

**Reply by sentriz** [reply]
which parameter is that? never heard of it

**Reply by sentriz** [reply]
theres no bandwidth issue, the client can request to get the transcoded stream or not, and the desired max bitrate

**Reply by Tolriq** [reply]
By default if I request format opus and maxbitrate 128kb Gonic will return me the raw FLAC file not what I ask.

This is a BW issue, the user configure the client to transcode and max birate but the server does not respect it and does not fail either.

https://opensubsonic.netlify.app/docs/endpoints/stream/

If you do not transcode to what is asked you should probably fail. But you should also transcode without forcing a profile server side. Client have the choice in the API.

**Reply by sentriz** [reply]
ah format, i see. how does the client know what availble "formats" are possible? gonic has many

![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/34270e60-e32c-4d83-a786-25636e486ec9)

**Reply by sentriz** [reply]
also, the most popular subsonic client, DSub, doesn't support format as far as I know. how have people been choosing different formats, if not on the server side? (and sorry to hijack your thread with this, @epoupon )

**Reply by dweymouth** [reply]
This seems like something else we should address in OpenSubsonic - clarifying what the format argument to `stream` actually supports. The official docs only give "mp3" or "flv" as examples, but we should perhaps clarify this. And a format is not just a simple thing either as it's actually a combo between a codec and a container (eg AAC over MPEG-4, which typically uses the .m4a extension)

**Reply by sentriz** [reply]
also, with Airsonic, you can create as many "formats" as you like on the server side. that's why I find it strange clients have been requesting custom formats. how do they know which are possible? and which client/servers support this?

**Reply by dweymouth** [reply]
To keep it simple, and since we don't need to deal with the complexities of video with many standard combos of container/codec, we could just document a list of formats and which servers support each. eg:

* "mp3" - standard MP3 audio
* "opus" - Opus audio in an OGG container
* "vorbis" - Vorbis audio in an OGG container
* "aac" - AAC-LC audio in a MPEG-4 container
* "flac" - FLAC audio in a FLAC container
* etc

This would be the backward-compatible way of doing it at least, but it would make it impossible for clients to request unusual formats (which is maybe desirable), like FLAC in an OGG container

**Reply by sentriz** [reply]
but as far as I remember, with airsonic, and probably the original subsonic server, you can specify a while bunch of different formats - each with their own shell pipelines 

it's plausible that you'd have 3 or 4 different pipelines for a single format (that's why with gonic you can have opus, opus with replay gain prenormailised, opus with a 6db boost. or you could even take midi or tracker files and render them to an audio file)

so I'm wondering how, originally even with sunsonic, the client would know what the possible formats are

**Reply by Tolriq** [reply]
Yes there's a full API for transcoding that needs to be done to be on par with other providers too and I'll work on that at some point. (Specially to only transcode unsupported formats and not flac for example, something many wants)

Most servers do support opus and mp3 as a parameter or fail if they don't recognize the format.

**But the main issue is still that clients can set a format (or not) and a maxBitrate. The server should respect those or fail or use a default transcoding.**

Returning the default file breaks the contract and will use a lot of network data from users that don't know they need to do something server side.

And they will probably blame the client that says hey I can transcode but transcode does not occur and the client is not even aware and in all case once they got their huge bill it's too late.

OpenSubsonic current changes and the ad I'll make about it by supporting it in next release will bring a lot of attention to the servers supporting the new stuff. Many people are seeking for a good music solution and replacement for Plex or Emby/Jellyfin for very large libraries.

So I hope you are ready for more users and associated support :p And can at least error if you do not know the format but not return the original file.

**Reply by Tolriq** [reply]
Ok so on my side all is working. Seems Gonic is already in prod version with those. @epoupon do you have some kind of ETA or plans ? Just to know if I adjust a little my planning to match or not.

**Reply by epoupon** [reply]
Well I was waiting for your tests to catch as many issues as possible :)
I will release it now then (3.41.0)

**Reply by Tolriq** [reply]
I did some quick tests and my code respect the API if the answers does not match I see it.

But I did not check that the actual values your returns are correct, (they seems to be correct) but you might want to to test on a private server with more data.

I can send you a test Symonium version if wanted.

**Reply by sentriz** [reply]
@Tolriq I agree there is a contact that can be broken if there are no server side  settings. but still I'm wondering - where has this format parameter come from?

 when I started with gonic my main client was DSub, still the oldest active subsonic client, and it has no mention of a format param. how do DSub users do this?

**Reply by Tolriq** [reply]
I don't use that app, and it does not really matter here. The issue is about maxBitrate.

If it's set then the server must respect it or fail it must not stream something that will consume maybe 10 times more mobile data without the client knowing.

We can discuss a proper API for transcoding, but the current one should still prevent unwanted network usage.

Edit: So you are free to continue to ignore the format for now, just be sure to handle maxBitrate.

**Reply by spl0k** [reply]
Some clients don't event know what to do with that `maxBitrate` parameter. I don't remember which client it was but I think I saw one making range requests on the `stream` endpoint each time with a different `maxBitrate` value, I guess I was streaming a VBR MP3 at that time.

**Reply by Tolriq** [reply]
The fact that a client is broken does not means that you should not properly handle the proper clients to avoid big issues.

What's worse a client that does not know how to use the API and have errors, or a user that get a 200$ phone bill because the server ignored the parameter the client sent?

It is always better to return an error when you can't fulfill a request than doing something unexpected that can have huge impact on the users.

Many people have limited data plans on mobile, and still quite a few even when on wifi on their home plans.

**Reply by epoupon** [reply]
@Tolriq updated server with a fake Iron maiden's album (was actually very easy to do with picard). This showcases disc titles + subroles for performers + multi release types
In next LMS release I will work on adding display artist names for songs/albums

**Reply by epoupon** [reply]
And just added as well another album with multiple album artists

**Reply by Tolriq** [reply]
@epoupon all works in Symfonium (except disc title that I do not yet support) Nice work. Will need to test lms with my large library now :(

**Reply by epoupon** [reply]
Well that was quite straightforward as I just exposed already handled stuff. Now new stuff will really need more work...

**Reply by epoupon** [reply]
I updated the os demo server with display name support.
Was quite straightforward for the OS API, but needed some hacks in the UI to recreate links inside the display names.

Ex for this album:
![image](https://github.com/opensubsonic/open-subsonic-api/assets/8963963/ed27a5e6-bb51-4d44-b76b-82bf0682fe59)
Display artist is "Au5 feat. Danyka Nadeau", album artists are "Au5" and "Danyka Nadeau"

Another example at track level:
![image](https://github.com/opensubsonic/open-subsonic-api/assets/8963963/6b018a5c-357c-40a3-9628-7a7cec04ae21)
Display artist is "Uncle-X + Dixan" and track artists are "Uncle-X " and "Dixan"

**Reply by Tolriq** [reply]
Do you use MB data for the that or hardcore the feat. or the +.

**Reply by epoupon** [reply]
I try not to depend on MB database (I consider tags should be enough).
For this, I just try to replace all artists in the display name string. If there is one single artist missing, I fallback to a list of artists (as it is in the current LMS release).
It seems to work in many cases, but the artist names may not match what is set in the string. 
![image](https://github.com/opensubsonic/open-subsonic-api/assets/8963963/9ee4e8c6-2cd9-4669-9e18-53567b4261a2) (here there are the same, but you get the idea)

**Reply by Tolriq** [reply]
Anyway just synced and it works :p

![Screenshot_20231014-154631](https://github.com/opensubsonic/open-subsonic-api/assets/932609/088cbf25-7fcd-40a9-9906-c41a4cf45bec)
![Screenshot_20231014-155105](https://github.com/opensubsonic/open-subsonic-api/assets/932609/8b0b24f7-3b2c-411a-b77b-0e379cbb8938)

**Reply by epoupon** [reply]
Nice!
By the way, not sure why we have the display album name at the song level, since it is alredy set at album level? Do we really need to repeat the info for each song?

**Reply by Tolriq** [reply]
You mean displayalbumartist? Because it's tied to the album artist on the song too. When you display the song info you can display that field without querying the album.

**Reply by Tolriq** [reply]
@epoupon Sent you a message on Symfonium forums don't know if you have notifications active there.

**Reply by epoupon** [reply]
Updated first discussion post for version v3.42.0

**Comment by dweymouth** [comment]
For Supersonic (client), status for 0.6.0 release:

* Multi-valued album artists
* Multi-valued album genres
* Multi-valued track artists

In 0.9.0 (upcoming):
* Show release types for album
* Comment column in tracklist (toggleable)

I've tested with @epoupon's LMS demo server and all looks good. @sentriz if you have your personal library set up for multi-valued tags feel free to give it a spin by building from the `feature/multi-artist-genre` branch. ~~Planning on releasing this probably tomorrow as 0.6.0.~~ ... and it's released!

**Comment by jeffvli** [comment]
For [Feishin](https://github.com/jeffvli/feishin) (client), status as of 0.4.1:

OpenSubsonic support is not implemented yet. Feishin still only supports Navidrome's internal API and Jellyfin API.

There are a few barriers preventing me from fully implementing OpenSubsonic in my app:

- [getSongList2 endpoint](https://github.com/opensubsonic/open-subsonic-api/discussions/18) -- allows for a standardized way to fetch songs instead of having to use `search2` or `search3`
- [Sort order for list endpoints](https://github.com/opensubsonic/open-subsonic-api/discussions/19) -- allows for asc/desc on list endpoints
- [Total record count for list endpoints](https://github.com/opensubsonic/open-subsonic-api/discussions/16) -- allows for clients to determine the actual page counts instead of having to infinitely fetch
- (Optional) Additional `search` query parameter for getAlbumList2 and other list endpoints to search for entities matching the query (same functionality as `search3` except only returns from the selected entity)

**Reply by Tolriq** [reply]
search3 can already only return songs or artists or albums if you pass the proper counts.

**Reply by jeffvli** [reply]
It does, but it doesn't provide any server-side sorting functionality.

**Reply by Tolriq** [reply]
IMO we need to think about new endpoints for the 3 types please also post in https://github.com/opensubsonic/open-subsonic-api/discussions/57 if you have sort / filter needs related to the new fields too.

**Comment by lachlan-00** [comment]
I'll be implementing all the new data in Ampache 6.2. I'm just finishing 6.1 now and will move onto subsonic focus for the next major

**Comment by metalheim** [comment]
[navidrome](https://github.com/navidrome/navidrome) implemented 
*as of 0.50.0*:
- musicBrainzId
- bpm
- comment

*as of 0.50.1*:
- genres for each album and song

**Comment by deluan** [comment]
Navidrome list is being maintained here: https://github.com/navidrome/navidrome/issues/2695

This is the current list, not sure if I'll be able to maintain the list in two places in the future :)

## Extensions
- getOpenSubsonicExtensions
- stream
- search3
- transcodeOffset
- formPost
- songLyrics (soon: https://github.com/navidrome/navidrome/pull/2656)

## New Fields

### Child
- played
- bpm
- comment
- sortName
- mediaType
- musicBrainzId
- genres
- replayGain

### AlbumID3
- played
- userRating
- musicBrainzId
- genres
- sortName
- isCompilation
- discTitles

### ArtistID3
- musicBrainzId
- sortName

**Reply by Tolriq** [reply]
@deluan links to a maintained page is ok too, as long as we can show to other clients and servers that this is growing and they should join :)
